File: roadhelp_app/lib/models/ward.dart
import 'dart:convert';

import '/models/province.dart';
import 'base_model.dart';
import 'district.dart';
import 'garage.dart';

class Ward extends BaseModel {
  //
  int? provinceId;
  int? districtId;

  //
  String? name;
  String? prefix;

  //Relationship
  List<Garage>? garages;
  Province? province;
  District? district;

  Ward({
    this.provinceId,
    this.districtId,
    this.name,
    this.prefix,
    //Relationship
    this.garages,
    this.province,
    this.district,
    //
    int? id,
    DateTime? createdAt,
    String? createdBy,
    DateTime? updatedAt,
    String? updatedBy,
    int? version,
    bool? deleted,
  }) : super(
          id: id,
          createdAt: createdAt,
          createdBy: createdBy,
          updatedAt: updatedAt,
          updatedBy: updatedBy,
          version: version,
          deleted: deleted,
        );

  //
  factory Ward.fromJson(Map<String, dynamic> json) {
    return Ward(
      provinceId: json['provinceId'],
      districtId: json['districtId'],
      name: json['name'],
      prefix: json['prefix'],
      //Relationship
      /*garages: Garage.fromJsonToList(json['garages']),
      province: Province.fromJson(json['province']),
      district: District.fromJson(json['district']),*/
      //
      id: json['id'],
      //createdAt: json['createdAt'],
      createdBy: json['createdBy'],
      //updatedAt: json['updatedAt'],
      updatedBy: json['updatedBy'],
      version: json['version'],
      deleted: json['deleted'],
    );
  }

  //
  static List<Ward> fromJsonToList(dynamic json) {
    return json.map<Ward>((element) => Ward.fromJson(element)).toList();
  }

  //
  String toJson() {
    return json.encode({
      'provinceId': provinceId,
      'districtId': districtId,
      'name': name,
      'prefix': prefix,
      //
      'id': id,
      //'createdAt': createdAt,
      'createdBy': createdBy,
      //'updatedAt': updatedAt,
      'updatedBy': updatedBy,
      'version': version,
      'deleted': deleted,
    });
  }
}

File: roadhelp_web/src/main/java/com/codedy/roadhelp/model/Ward.java
package com.codedy.roadhelp.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;

import javax.persistence.*;
import java.util.HashMap;
import java.util.List;

@Entity
@Table(name = "ward")
public class Ward extends BaseModel {

    //region - Define Fields -
    @Column(name = "_name")
    private String name;

    @Column(name = "_prefix")
    private String prefix;
    //endregion


    //region - Relationship -
    @JsonIgnore
    @OneToMany(mappedBy = "ward", fetch = FetchType.LAZY, cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.DETACH, CascadeType.REFRESH})
    private List<Garage> garages;

    @JsonIgnore
    @ManyToOne(cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.DETACH, CascadeType.REFRESH})
    @JoinColumn(name = "_province_id") //updatable = false, insertable = false
    private Province province;

    @JsonIgnore
    @ManyToOne(cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.DETACH, CascadeType.REFRESH})
    @JoinColumn(name = "_district_id") //updatable = false, insertable = false
    private District district;
    //endregion


    //region - Getter & Setter -
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getPrefix() {
        return prefix;
    }

    public void setPrefix(String prefix) {
        this.prefix = prefix;
    }

    public List<Garage> getGarages() {
        return garages;
    }

    public void setGarages(List<Garage> garages) {
        this.garages = garages;
    }

    public Province getProvince() {
        return province;
    }

    public void setProvince(Province province) {
        this.province = province;
    }

    public District getDistrict() {
        return district;
    }

    public void setDistrict(District district) {
        this.district = district;
    }
    //endregion


    //region - Relationship Helper -

    /**
     * Hàm này trả về cấu trúc nguyên thủy của entity này.<br/><br/>
     * <p>
     * Viết bởi: Hiếu iceTea<br/>
     * Ngày: 23-10-2021<br/>
     * Thời gian: 22:22<br/>
     *
     * @return
     */
    public HashMap<String, Object> toHashMap() {
        HashMap<String, Object> hashMap = super.toHashMap();

        hashMap.put("provinceId", province != null ? province.getId() : null);
        hashMap.put("districtId", district != null ? district.getId() : null);

        hashMap.put("name", name);
        hashMap.put("prefix", prefix);

        return hashMap;
    }

    @JsonProperty("garages")
    public List<HashMap<String, Object>> getGarageHashMap() {
        return garages != null ? garages.stream().map(Garage::toHashMap).toList() : null;
    }

    @JsonProperty("province")
    public HashMap<String, Object> getProvinceHashMap() {
        return province != null ? province.toHashMap() : null;
    }

    @JsonProperty("district")
    public HashMap<String, Object> getDistrictHashMap() {
        return district != null ? district.toHashMap() : null;
    }
    //endregion

}
